Cross compile

deitch opened the pull request:
## Description
- support cross-compile for `arm64` and `ppc64le` from `amd64`, in addition to native builds
- update `Dockerfile.<arch>` to use period instead of dash, to align with how `projectcalico/go-build` does it
- explicitly name file in `docker-images/bin`

Still missing: target to build all, will appear in a separate PR

Depends on https://github.com/projectcalico/go-build/pull/26
## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note

```release-note
Add support to cross-build felix for `arm64` and `ppc64le` from `amd64`, in addition to native (`amd64` to `amd64`; `arm64` to `arm64`; `ppc64le` to `ppc64le`) builds.
```

This is based to a large degree on the work done by @djlwilder .

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/projectcalico/felix?pullRequest=1703) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/projectcalico/felix?pullRequest=1703) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: djlwilder<br/>:x: deitch<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/projectcalico/felix?pullRequest=1703) it.</sub>

deitch commented:
Note: I didn't mark off the unit tests or integration tests one way or another. I didn't change any of the code, just the compile and image build stages, so no changes in tests.

deitch commented:
I hate having to figure out conflicts. Sigh.

deitch commented:
The `Makefile` changed under the covers on me. This is going to be messy....

deitch commented:
git is doing weird stuff here. I am going to try and reset. If not, I will just open a new PR instead of this one.

deitch commented:
Feh, that took forever. git kept bringing back old changes. Done.

deitch commented:
> We lost the definition of FV_FELIXIMAGE

Must have been that weird rebase. I kept completing the rebase, and it would ask me to fix the merge again, same issues I just had fixed. Will fix.

deitch commented:
> We lost the definition of FV_FELIXIMAGE

Done. Thanks for pointing it out.

deitch commented:
Also updated how we get the qemu binaries to match cross-compile. Truth is, I would prefer to just take the qemu binaries from the `go-build` image, since it is there. A different day...

djlwilder commented:
I see the problem with Ci in this PR. The scripts used for running fv on semaphoreci are running the test outside of the Makefile (@fasaxc).  Therefore FV_FELIXIMAGE is not set and the image name is defaulting to calico/felix:latest.  This PR changes the image name to calico/felix-amd64:latest so the test fails.  Once we start building multi-arch Docker images the image name should go back to calico/felix (I think).

djlwilder commented:
> FV_FELIXIMAGE is not set and the image name is defaulting to calico/felix:latest. 
I take that back,  it looks like FV_FELIXIMAGE is not ever used in the fv code,  so it always defaults calico/felix:latest.

deitch commented:
> it looks like FV_FELIXIMAGE is not ever used in the fv code, so it always defaults calico/felix:latest

Fixable easily enough. Will push shortly.

> This PR changes the image name to calico/felix-amd64:latest

I think it always was `calico/felix-${ARCH}:latest`. I don't like that style, prefer `${IMAGE_NAME}:${ARCH}-latest`. That creates a single repository for the images, with different versions (including different archs) separated by tag. Add 5 new archs or OSes, and you just need more tags, not more repos. It also makes the multi-arch manifests easier (and is what the Docker team does).

Nonetheless, I thought it was too much to push at once, so left it. Also why I didn't do cross-image (single make target to build and push everything) or multi-arch manifests. Figure get this all working as an option and merged, then single make target, then manifests. One PR at a time keeps people sane (well, at least me... relatively sane, anyways...).

deitch commented:
OK, I think I got it here @djlwilder . There were 3 explicit references to the image name in the `fv/` code: 2 in test files and 1 in the actual code. The one in the actual code references `utils.Config.FelixImage`, while the two in test files had it hard-coded (funny, one as `calico/felix`, the other as `calico/felix:latest`, which is the same thing, but still).

1. Updated the two test file references to use `utils.Config.FelixImage` as well, so one place to manage.
2. Set the environment variable in the `run-test` script to override it. This should make it easy to change in the future.

Back to you!

deitch commented:
And rebased to fix merge conflict

djlwilder commented:
Thank you @deitch for chasing the test issue.    I discovered that I was hitting the same problem on my my ppc64le native Ci.    Since this is a separate problem from this PR (and I needed a fix for ppc64) I pulled your changes into a separate PR.  

[](url) https://github.com/projectcalico/felix/pull/1708

I made a couple of additions from what you did.  I now propagate the value of FV_FELIXIMAGE  from the Makefile rather than set it in fv/run-test.  I also make the same change for the k8sfv/ tests.  If my PR is merged before this one then a re-base of your fork/branch should merge our changes.  Please comment on the new PR if you see any issues.

Unfortunately none of this will fix the ci failure for this PR :(  The ci scripts used here are not using the Makefile, run-test or or setting FV_FELIXIMAGE.

deitch commented:
> Thank you @deitch for chasing the test issue

And thank you for pointing me in the right direction. :-)

> If my PR is merged before this one then a re-base of your fork/branch should merge our changes

Works for me.

> The ci scripts used here are not using the Makefile, run-test or or setting FV_FELIXIMAGE and we changed the calico/felix image name

Who has admin access to the semaphore account used here? Let's get it to call the wrapper, and then whatever changes we do I the future (like multi-arch manifest) will be changeable with ease via env vars, not changing semaphore again.

Do you know who can fix that?

djlwilder commented:
>>    The ci scripts used here are not using the Makefile, run-test or or setting FV_FELIXIMAGE and we changed the calico/felix image name

> Who has admin access to the semaphore account used here? Let's get it to call the wrapper, and then whatever changes we do I the future (like multi-arch manifest) will be changeable with ease via env vars, not changing semaphore again.


@fasaxc might help here.  In the past ci used make fv,  probably should go back to that.

djlwilder commented:
@deitch  I have not looked at how to create a multi-arch manifest image yet.  Is that the next step?  At this point we can create 3 images,  calico/felix:<amd64,arm64,ppc64le>-<version>.  Would the multi-arch stuff then give us a single artefact called  calico/felix:<version> ?  That would fit well with the current release process.


 BTW:  the docker/moby PR you mentioned to do the multi-arch stuff was merged!

deitch commented:
Yes that is it. I have done quite a few in my work with linuxkit. 

After we get this in, we can create another PR for single makefile target that builds all platforms, then we will do another for multiarch manifest. 

I saw it was merged in. Much easier that way!

deitch commented:
OK, so next steps, as far as I can tell:

1. Fix semaphore (@fasaxc ?)
2. Merge this one in
3. New PR with build target that builds all archs and all docker images
4. New PR that does the above with multi-arch manifest

Is that right? How do we move it ahead?

caseydavenport commented:
Just an FYI @fasaxc is out of the office until sometime next week.

djlwilder commented:
Thanks @caseydavenport 
@deitch I plan to look at this more tomorrow.  Thanks.

fasaxc commented:
@deitch I own the SemaphoreCI project.  I recently "unrolled" the test logic as an experiment to split it into parallel jobs.  I've been planning to roll is back into `make fv` so the the Semaphore invocation would be something like `make fv GINKGO_BATCH=3 GINKGO_NUM_BATCHES=7` and the makefile would handle the actual fan out.

deitch commented:
Hi Shaun, great to (e)meet you. Curious how "Shaun" became handle "fasaxc". :-)

> I own the SemaphoreCI project

Great! What can we do to get this moving ahead? Really looking forward to seeing Calico truly cross-platform, as is @djlwilder who has been putting work in as well.

fasaxc commented:
Likewise.  Haha, `fasaxc` was my (teenage) effort to make up a username that I wouldn't have to share ;-)

A quick skim on the PR looks like it could be a really nice approach.  Are the fv's passing for you if you run `make fv`?  If so, I'll try and get that fix to roll the test incantation back into the makefile in over the next few days so that we can get this moving.

deitch commented:
> fasaxc was my (teenage) effort to make up a username that I wouldn't have to share

I have a unique enough last name, that I just use it, or a shortened form. I even own deitcher.net. Missed deitcher.com by a few months. 

No chance I am going to try to grab `fasaxc` :-)

> Are the fv's passing for you if you run make fv

I think so, but it has been a while since I ran it - so many projects, so little time - so I will rerun, rather than rely on my faulty memory.

One thing, though, is that it does depend on [projectcalico/go-build#26](https://github.com/projectcalico/go-build/pull/26) , which is the cross-compile for the basic `go-build`. So I can run locally, since it all is here, but we need that one in first for everything to work.

deitch commented:
> Are the fv's passing for you if you run make fv

Ok, definitely failing, but probably because I am running on my Mac:

```

fasaxc commented:
The go-build change looked good so I merged that.  Ah, but Semaphore is expecting the wrong image name.  I'll sync that up...

deitch commented:
> Ah, but Semaphore is expecting the wrong image name

Yeah, we tried to clean up the images too, so it is a single image repo but multiple tags. That is the way we do the multi-arch for all the official library images (I work on linuxkit as well). Once these are in, and everything is working well, we can do multi-arch manifests. The docker CLI finally merged it in a week or so ago.

fasaxc commented:
I've synced that up now, the new go-build is pushed as `calico/go-build:v0.10` (and `calico/go-build:v0.10-amd64`).  Semaphore is now using the `make push` target

deitch commented:
> Semaphore is now using the make push target

keeps getting better and better. Why is semaphore failing now?

fasaxc commented:
> Why is semaphore failing now?
Which job?  The felix one?  (I haven't fixed up the felix one yet; not sure if I'll be able to get to that today.)

deitch commented:
> Which job? The felix one?

Yeah, that was what I meant. When you wrote:

> Semaphore is now using...

I thought you meant you had cleaned up _this_ Semaphore too.

Great, thanks.

djlwilder commented:
I have successfully run fv with these changes,  I just ran it again to verify.

Run on amd64 (linux), all tests passed.
```
make calico/felix
ARCH=ppc64le make calico/felix
ARCH=arm64 make calico/felix

$ docker images | grep felix
calico/felix-arm64                         latest                     261b9347e98a        13 seconds ago       69.4MB
calico/felix-ppc64le                       latest                     7b41bdf69567        About a minute ago   80.4MB
calico/felix-amd64                         latest                     5b1241d730a2        2 minutes ago        66.3MB

make ut
make fv
JUST_A_MINUTE=true CLEANUP=true USE_TYPHA=false make k8sfv-test
JUST_A_MINUTE=true CLEANUP=true USE_TYPHA=true make k8sfv-test
```
I was able to run the test currently failing on semaphore after adding the export FV_FELIXIMAGE line.
```
make calico/felix bin/iptables-locker bin/test-workload bin/test-connection fv/fv.test
cd fv
export FV_FELIXIMAGE=calico/felix-amd64:latest
export FV_TYPHAIMAGE=calico/typha:v0.6.0-beta1-16-g512a0f2
export FV_ETCDIMAGE=quay.io/coreos/etcd:v3.2.5
export FV_K8SIMAGE=gcr.io/google_containers/hyperkube:v1.7.5
./fv.test -ginkgo.parallel.node 1 -ginkgo.parallel.total 7 -ginkgo.seed 1 -ginkgo.randomizeAllSpecs -ginkgo.slowSpecThreshold 90
```

djlwilder commented:
BTW: @fasaxc with these changes "make calico/felix" (on amd64) produces an images named calico/felix-amd64, previously the image was named calico/felix.  Once we add the multi-arch image code it will produce a multi-arch-image named calico/felix again.  In the interim the release target will likely be broken.

deitch commented:
> I was able to run the test currently failing on semaphore

Thanks @djlwilder . I probably could have just spun up a vm or grabbed a digital ocean droplet or EC2 instance or packet host, but my mind wasn't thinking straight. I appreciate your stepping in.

fasaxc commented:
@deitch @djlwilder I'd rather not break the release target (and the downstream builds).  Until we have a multi-arch image, how about simply retagging the amd64 image as `calico/felix` until we have a multi-arch `calico/felix` image?

I think it'll be worth tweaking the release target and downstream builds to use the multi-arch image, but it's a fair amount of work to plumb through a new image (and keep older releases building too).

deitch commented:
@fasaxc are you in the UK? I didn't think you would be commenting on this at that hour from CA. Part of the original Tigera team was based on Metaswitch staff, which is a UK company. Are you from there?

> I'd rather not break the release target (and the downstream builds). Until we have a multi-arch image, how about simply retagging the amd64 image as calico/Felix?

Sure. I can see how it is necessary, until everything is ready. We could make this multi-arch manifest, but I really would like to keep away from that complexity until we have all of the builds working.

I also notice the "docker push" comments are somewhat out of date. I will fix those comments in the `Makefile`.

This PR does all the hard work of making local compiles and cross-compiles possible. After this one is done, I will add a target for compiling _all_ from `amd64`. Then we can do multi-arch manifest after those steps are done for all.

deitch commented:
Hmm, why are we using `$(ARCH)` in the image name and not the tag? We cleaned that up in `go-build`, but not here. 

(I ask the question of myself, since I wrote this PR)

I assume it was so as not to break downstream builds, which may depend on these tags. But then again, we weren't doing much with the tags. I am going to change them as a separate commit, so we always can revert. First a commit to create the default, then one to change the `ARCH` from the image name to the tag.

deitch commented:
> Until we have a multi-arch image, how about simply retagging the amd64 image as calico/Felix?

Done.

deitch commented:
I am going to put in the other commit now, to have a single repo `calico/felix`, but arch by tag.

deitch commented:
Oh, I definitely messed up the `Makefile`. Hold on, updating the commit....

deitch commented:
OK, now I got it. I really messed up with the tags there, but now it looks right. I think....

@fasaxc like to get your feedback there. I fixed the image tagging, fixed the various parts of the `Makefile` related to `docker tag` and `docker push` and comments, and created a "noarch" default for `amd64`

djlwilder commented:
Latest changes LGTM.  @fasaxc "make fv" is using the updated image name.  Semaphore needs to be update as well:
  "export FV_FELIXIMAGE=calico/felix:latest-amd64"   <<<  new image name.

fasaxc commented:
@deitch please can you make sure that the amd64 image also gets tagged without a suffix (i.e. as `calico/node:latest`)?  There are "downstream" builds, such as `calico/node` that are expecting the old name.

Once `calico/node`'s build is updated to multi-arch then the `calico/felix:latest` name can be phased out (or repurposed as the multi-arch image).

deitch commented:
@fasaxc I don't get it. The image name `calico/node` doesn't appear anywhere in here. Is this something new we need?

deitch commented:
Drat. I need to rebase. Did I miss a merge in the last 24 hours?

fasaxc commented:
@deitch `calico/node` is built from the `projectcalico/calico` repo, one of the inputs to that build is felix, which is gets from the `calico/felix` image.  Doing it that way lets us pass our artefacts from this build to that in a reasonably clean way while using public CI infrastructure.

deitch commented:
@fasaxc now I am even more confused. :-) You wrote:

> can you make sure that the amd64 image also gets tagged without a suffix (i.e. as calico/node:latest)

But no such image existed before in the `Makefile` anywhere. I am fine adding a new image tag, if it is needed, but I don't understand where it came from.

deitch commented:
This rebase is going horribly wrong. I cannot even figure out what changed to rebase it.

deitch commented:
> This rebase is going horribly wrong. I cannot even figure out what changed to rebase it.

I had to redo the change manually. Wiped out the old branch, new `checkout -b cross-compile`, reapply changes. What a mess. But I finally got it.

When git misbehaves, it *really* misbehaves.

deitch commented:
So back to the `calico/node`, @fasaxc . I am happy to do what you want here, just didn't understand it. Is this something totally new? There definitely was no `calico/node` anywhere in the `Makefile` before.

djlwilder commented:
@fasaxc  the release-once-tagged: target is tagging the image as you suggested.
docker tag calico/felix:latest-amd64 calico/felix:latest
Is that not sufficient?

fasaxc commented:
@deitch I'm talking about this repo: https://github.com/projectcalico/calico/tree/master/calico_node

The `calico/felix` image only contains Felix, but to run Calico on a host, you need other components too.  `calico/node` is the main container image that we ship.  It contains Felix, but also BIRD, confd, configuration templates, startup logic (to initialise the datastore before starting Felix/BIRD/etc).

So that our CI can be mostly out in the open, the way we get the `calico-felix` binary from [Felix's semaphore build](https://semaphoreci.com/calico/felix-2) to [`calico/node`'s semaphore build](https://semaphoreci.com/calico/calico) is that the Felix one does a `make calico/felix` and then tags it with  `docker tag calico/felix:latest calico/felix:$BRANCH_NAME` and pushes it.  Then the `master` build of calico node pulls that image and extracts `calico-felix` from it.  

If we change the names of binaries/images that process will be broken. We absolutely could tweak the semaphore builds to expect the new image name and retag like this:`docker tag calico/felix:latest-amd64 calico/felix:$BRANCH_NAME` but then semaphore won't be able to build any branch apart from master (or we'll have to add conditional logic to semaphore, which gets old fast).

@djlwilder The `make release` target is only used when we do a versioned release.  Its purpose is to cut the git tag, do a clean build with the correct version information in the binary and retag the images with the correct names.

deitch commented:
> the way we get the calico-felix binary from Felix's semaphore build to calico/node's semaphore build is that the Felix one does a make calico/felix and then tags it with docker tag calico/felix:latest calico/felix:$BRANCH_NAME and pushes it. Then the master build of calico node pulls that image and extracts calico-felix from it.

Ah, ok, now I get it.

We already do that. From the `Makefile` in this PR:

```Makefile
	docker tag calico/felix:latest-$(ARCH) calico/felix:latest
	docker tag calico/felix:latest-$(ARCH) quay.io/calico/felix:latest
	docker tag calico/felix:latest calico/felix:$(VERSION)
	docker tag calico/felix:$(VERSION) quay.io/calico/felix:$(VERSION)
```

Net result is that you have an image for `amd64` named `calico/felix:latest` and another `calico/Felix:$(VERSION)` available, which should meet the requirement.

Also, from `Dockerfile.amd64`:

```Dockerfile
ADD bin/calico-felix-amd64 /code/calico-felix
```

So in the image, the file is named `calico-felix`.

These default "arch-less" tags were done specifically as you requested it. 

Does this meet the requirements?

deitch commented:
Speaking of which:

> the Felix one does a make calico/felix and then tags it with docker tag calico/felix:latest calico/felix:$BRANCH_NAME and pushes it

Where is that push? I don't see any `docker push` anywhere in the `Makefile` except in documentation. Is it semaphore doing it?

fasaxc commented:
FYI the Makefile patch to allow Semaphore to be fixed up is in progress under #1732

fasaxc commented:
@deitch With the change in Semaphore, the main FV tests are now passing again (although they'll run really slowly until your branch is rebased onto #1732) but the `k8sfv` tests are not.  Please can you take a look?

deitch commented:
@fasaxc :

> until your branch is rebased onto #1732

I cannot rebase until #1732 is merged in, though.

> oh no, I need to rebase again? Every time I need to rebase this, it gets caught up in weird file conflicts. Oh well.

> but the k8sfv tests are not. Please can you take a look?

Surely!

deitch commented:
Phew. That rebase was much less painful than expected. @fasaxc we really need to get this merged in. The rebases are doing me in. I will look at that remaining `k8sfv` issue.

deitch commented:
OK @fasaxc got it. It was missing an env var setting for `k8xfv-test` target. Repushed. Hopefully semaphore will go through now. I will keep an eye.

deitch commented:
Ok @fasaxc all of the `k8sfv` tests passed. Looks like the only thing that failed in Semaphore was 3 of the 217 tests in `Run FV tests batch 1`, and it looks like a latency issue?

```
Summarizing 3 Failures:

[Fail] Latency tests with initialized Felix and etcd datastore IPv4: Network sets tests with initialized Felix and etcd datastore [It] with allow-all should have good latency 
/go/src/github.com/projectcalico/felix/fv/latency_test.go:167

[Fail] Latency tests with initialized Felix and etcd datastore IPv4: Network sets tests with initialized Felix and etcd datastore with all() source selector [It] should have good latency 
/go/src/github.com/projectcalico/felix/fv/latency_test.go:183

[Fail] Latency tests with initialized Felix and etcd datastore IPv4: Network sets tests with initialized Felix and etcd datastore with all() source selector with 10k IPs in an IP set [It] should have good latency 
/go/src/github.com/projectcalico/felix/fv/latency_test.go:210
```

I suspect that would pass if run again?

deitch commented:
Yes! It all passed @fasaxc ! Can we merge? (as in, before I have to rebase again, please....)

fasaxc commented:
@deitch Looks like we've now got a conflict due to the fix-up that I did for semaphore.  You need to rebase onto that (or merge it in).

deitch commented:
Rebased, CI is running. @fasaxc . We really should merge it in when it passes. :-)

deitch commented:
Yes! CI passed @fasaxc ! Merge?

fasaxc commented:
Sorry for the delay, I want to give this a try locally and make sure that our release/downstream builds process will work as normal before merging.

deitch commented:
Great. Let me know if you find anything and I’ll address. I’m on holiday through end of the weekend, so addressing anything will be next week. 

Once this is in, next plan is to make a simple target to build all via cross build and use manifest-tool for single image multi arch. Then I’ll adfress the rest of the downstream components

fasaxc commented on Makefile:
```diff
@@ -42,40 +42,55 @@
-FV_FELIXIMAGE?=$(FELIX_IMAGE_NAME)
-FV_ETCDIMAGE?=quay.io/coreos/etcd:v3.2.5$(ARCHTAG)
-FV_K8SIMAGE?=gcr.io/google_containers/hyperkube$(ARCHTAG):v1.7.5
+GO_BUILD_VER ?= latest
```
We need to get a pinned version in here.  I believe `v0.10` is the current `latest`

deitch replied:
It is `v0.10` (I just checked), and that is a good idea. Fixing.

fasaxc commented on Makefile:
```diff
@@ -42,40 +42,55 @@
-	PROTOC_VER?=v0.1
-	FV_TYPHAIMAGE?=calico/typha:v0.6.0-beta1-16-g512a0f2
-endif
+# Both native and cross architecture builds are supported.
```
Please can you list the known-supported ARCHes?

deitch replied:
Basically, any one in `docker-images/Dockerfile.*`. I will use the code from the `go-build` Makefile to list them out.

deitch replied:
Done. It adds it to the output of `make help`.

deitch replied:
Actually, I added a little more instructional help there.

deitch commented on Makefile:
```diff
@@ -600,8 +651,12 @@ release-once-tagged:
 	@echo "Felix release artifacts have been built:"
 	@echo
 	@echo "- Binary:                 bin/calico-felix"
```
Good point. Fixing.

deitch commented on Makefile:
```diff
@@ -580,13 +624,20 @@ release-once-tagged:
 	@echo
 	$(MAKE) bin/calico-felix calico/felix
-	docker tag calico/felix:latest quay.io/calico/felix:latest
+	# default image until we use multi-arch manifest
```
> Does running make release-once-tagged ARCH=ppc64le work, for example?

It does. More correctly, it runs all of the correct targets with the correct vars.

deitch replied:
`make release` should be fine. All it does is change the release notes and add them as a tag, which has no impact on the compiled binaries or images.

deitch replied:
> Given how many images there now are, it might be good to have a make push VERSION=3.0.0 target to do all the pushing

Yes. I didn't know why it didn't have one. I assumed there was some reason to which I wasn't to privy. Separate PR though. The rebasing has been so much work with each change, and this one open for so long, we should get it in, then do a simple PR for that.

fasaxc replied:
Agreed

deitch commented:
> I don't think the make release issue is a blocked for merging but we'll need a solution if we're actually going to start releasing the non-amd64 images after this work.

I had planned on creating a multi-arch target that would wrap and build them all.

deitch commented on Makefile:
```diff
@@ -191,16 +229,19 @@ calico-build/centos6:
-	cp bin/calico-felix docker-image/bin/
-	docker build --pull -t $(FELIX_IMAGE_NAME) --file ./docker-image/Dockerfile$(ARCHTAG) docker-image
+	cp bin/calico-felix-$(ARCH) docker-image/bin/
+	docker build --pull -t calico/felix:latest-$(ARCH) --file ./docker-image/Dockerfile.$(ARCH) docker-image
```
I read this comment through several times, still don't get it. We cannot run this as `docker build --pull -t calico/felix:latest ... ` (as opposed to `calico/felix:latest-$(ARCH)` ), since it is building an arch-specific image. If we do, we can trounce other images.

Or are you saying that in _addition_ to the `ARCH`-specific tag to provide a `calico/felix:latest` tag if `ARCH=amd64`? We do that on [LL 638-643](https://github.com/projectcalico/felix/pull/1703/files#diff-b67911656ef5d18c4ae36cb6741b7965R638)

fasaxc replied:
I mean "in addition".  Lines 638-643 are in the wrong target for our CI systems.

For context, our SemaphoreCI script looks like a bit like this:
```
make calico/felix
docker tag calico/felix:latest calico/felix:`git describe`
docker push  calico/felix:`git describe`
```

If you do the retagging in the release-once-tagged target then I think the Semaphore script will break.  I'd like to keep the old names for now, until we've got the whole pipeline ready for multi-arch, then we only need to update Semaphore once.

Updating Semaphore is a bit painful because ideally, we need it to keep building old release branches too.  That means that every change needs to be put behind version detection logic, which is a pain.

I repushed your branch to the main repo so that the above logic will run (it's normally blocked for PRs) and it does fail: https://semaphoreci.com/calico/felix-2/branches/smc-multi-arch/builds/1

deitch replied:
> I mean "in addition". Lines 638-643 are in the wrong target for our CI systems.

Ah, ok, now I got it.

It is for reasons like that that almost all of the CI work I do is just `make <something>`. This lets me control everything from the `Makefile`, rather than worrying that something is tied too tightly into a specific command in CI.

OK, I will add a conditional `docker tag` there right away. 

At some point really soon, we are going to have to change Semaphore to do multi-arch builds with manifests. Actually, as soon as this PR is in, I will open another for that. 

Is there any chance I can convince you to change this:

```
make calico/felix
docker tag calico/felix:latest calico/felix:`git describe`
docker push  calico/felix:`git describe`
```

to something like:

```
make calico/felix
make tag
make push
```

And then I will add the `tag` and `push` targets?

deitch replied:
Just added another commit with:

```makefile
	ifeq ($(ARCH),amd64)
	docker tag calico/felix:latest-$(ARCH) calico/felix:latest
	endif
```

fasaxc replied:
> Is there any chance I can convince you to change this:
> 
> make calico/felix
> docker tag calico/felix:latest calico/felix:`git describe`
> docker push  calico/felix:`git describe`

Yes, I'm expecting we'll need to change it at some point.  I just don't want to have to change it more than once if I can help it so that we don't accumulate lots of conditional cruft.

Some of our build have a `make semaphore` target, which works OK.  However, felix's Semaphore build is optimised to run FV tests in lots of batches, which makes that difficult.

deitch replied:
> I just don't want to have to change it more than once if I can help it so that we don't accumulate lots of conditional cruft.

That makes sense. I have been working through this a little blindly since, unlike with travis or circle or Jenkins, I cannot see the CI commands we use for Semaphore.

Either way, I got the extra tagging in you requested in the right place. It looks like Semaphore either won't run or won't report until they are approved.

Once this PR is through, I will open the new one for "single make target builds all arch images", which will require those changes, but you can walk me through getting it adapted to Semaphore correctly then.

fasaxc replied:
> That makes sense. I have been working through this a little blindly since, unlike with travis or circle or Jenkins, I cannot see the CI commands we use for Semaphore

If you go to a specific run (for example, this master run: https://semaphoreci.com/calico/felix-2/branches/master/builds/921) and expand each job, you should be able to see the individual commands.

deitch commented:
OK, I think I addressed all of your comments (except for the one with the open question). CI is running, hopefully last time? :-)

fasaxc commented:
Build passed here: https://semaphoreci.com/calico/felix-2/branches/master/builds/921

fasaxc commented:
@deitch if you're planning to suggest some new makefile targets, I suggest you get on our slack in the calico-dev channel.  I know @caseydavenport is looking to streamline and standardise our targets so let's make sure we sync up with that effort.

deitch commented:
It felt *so* good to do `git br -d cross-compile` ! :-) Thanks for all of the help @fasaxc .

deitch commented:
> I suggest you get on our slack in the calico-dev channel

OK, will do. I will be happy to speak with @caseydavenport there. I do have Slack overload, but it is the price we pay. It is better when I set notifications to show just when I get a DM or am mentioned.

> if you're planning to suggest some new makefile targets

I am. We now can build for any given arch right here. Next step is to build for **all** archs when on `amd64`, push out all of the images, and create a multi-arch manifest, so that `docker pull calico/felix` will pull the correct arch-specific one.

chakravarthy392 commented:
Hi, this also support cross compilation for s390x right ?!

deitch commented:
@chakravarthy392 we had some issues with cross-compilation for s390x. Actually, we didn't have issues with cross-compilation, we had issues relating to `qemu` and building images. My understanding is that `qemu` virtualization isn't officially support for s390x, but I could be remembering incorrectly. 

See [here](https://github.com/projectcalico/felix/blob/master/Makefile#L50-L53)

chakravarthy392 commented:
Thanks for the quick response @deitch , I will check that `qemu` issue.
If you don't mind, I have one more question.
Currently I am building Felix on s390x architecture and I want to raise a PR to pin protoc version to latest , but I can't do it w.r.t master as s390x is tag is removed from the code. So, can I raise a PR w.r.t `release-v3.1` branch?

deitch commented:
I don't know what you mean by:

>  s390x is tag is removed from the code
